packages/backend-common/src/database/SingleConnection.ts

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Knex from 'knex';
+import { Config } from '@backstage/config';
+import { createDatabaseClient, ensureDatabaseExists } from './connection';
+import { PluginDatabaseClientFactory } from './types';
+
+/**
+ * Implements a Database Manager which will automatically create new databases for plugins when requested. All
+ * requested databases are created with the credentials provided.
+ */
+export class SingleConnectionManager {
```

Comment:
Thinking we should toss in the term `database` here so it's more clear and to my note. It might also be a bit simpler to just combine the plugin factory method into this class. Might look something like this:

```ts
export class SingleConnectionDatabaseManager implements PluginDatabaseManager {
   async fromConfig(config: Config) {
    // create instance
  }

  // This would be the simple interface where a plugin request a database connection
  getDatabaseClient(pluginId: string): knex {
    // ensureDb
    // return new knex client
  }

  // if we wanted to restrict access from plugins using other databases then something like this could be used
  forPlugin(pluginId: string): PluginDatabaseManager {
    // create a wrapper that adheres to a restricted PluginDatabaseManager interface.
  }
}
```

Reply:
I can do the renaming, and to also define a fromConfig method for consistency. I didn't know that was the earlier pattern.

I'm not sure what you mean by combining the plugin factory method though, isn't this one class already? If the objective is to give the plugin the entire object, we'd need to change the `PluginDatabaseClientFactory` type to now be an interface, and that breaks the enforcement that different plugins get different databases (because plugins can just call getClient -- and the argument to getClient should actually be `getClient(database: string)`).

Reply:
Addressed in https://github.com/spotify/backstage/pull/2697/commits/35c0294018b8433787a795b7d9fcb6e9daef2b68.